File: 3625.count-number-of-trapezoids-ii.cpp
#
# @lc app=leetcode id=3625 lang=cpp
#
# [3625] Count Number of Trapezoids II
#
# @lc code=start
class Solution {
public:
    int countTrapezoids(vector<vector<int>>& points) {
        int n = points.size();
        int count = 0;
        
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                for (int k = j + 1; k < n; k++) {
                    for (int l = k + 1; l < n; l++) {
                        if (isTrapezoid(points[i], points[j], points[k], points[l])) {
                            count++;
                        }
                    }
                }
            }
        }
        
        return count;
    }
    
private:
    bool isTrapezoid(vector<int>& p1, vector<int>& p2, vector<int>& p3, vector<int>& p4) {
        vector<vector<int>*> pts = {&p1, &p2, &p3, &p4};
        
        vector<vector<int>> orderings = {
            {0,1,2,3}, {0,1,3,2}, {0,2,1,3}, {0,2,3,1}, {0,3,1,2}, {0,3,2,1}
        };
        
        for (auto& ord : orderings) {
            if (isValidTrapezoid(pts[ord[0]], pts[ord[1]], pts[ord[2]], pts[ord[3]])) {
                return true;
            }
        }
        
        return false;
    }
    
    bool isValidTrapezoid(vector<int>* a, vector<int>* b, vector<int>* c, vector<int>* d) {
        // First verify convexity (R2)
        if (!isConvex(a, b, c, d)) {
            return false;
        }
        
        // Then check parallel sides (R3)
        bool parallel1 = isParallel(*a, *b, *c, *d);
        bool parallel2 = isParallel(*b, *c, *d, *a);
        
        return parallel1 || parallel2;
    }
    
    bool isConvex(vector<int>* a, vector<int>* b, vector<int>* c, vector<int>* d) {
        // Check if quadrilateral ABCD is convex using cross products
        auto cross = [](vector<int>* p1, vector<int>* p2, vector<int>* p3) {
            long long dx1 = (*p2)[0] - (*p1)[0];
            long long dy1 = (*p2)[1] - (*p1)[1];
            long long dx2 = (*p3)[0] - (*p2)[0];
            long long dy2 = (*p3)[1] - (*p2)[1];
            return dx1 * dy2 - dy1 * dx2;
        };
        
        long long c1 = cross(a, b, c);
        long long c2 = cross(b, c, d);
        long long c3 = cross(c, d, a);
        long long c4 = cross(d, a, b);
        
        return (c1 > 0 && c2 > 0 && c3 > 0 && c4 > 0) || 
               (c1 < 0 && c2 < 0 && c3 < 0 && c4 < 0);
    }
    
    bool isParallel(vector<int>& a, vector<int>& b, vector<int>& c, vector<int>& d) {
        long long dx1 = b[0] - a[0];
        long long dy1 = b[1] - a[1];
        long long dx2 = d[0] - c[0];
        long long dy2 = d[1] - c[1];
        
        return dy1 * dx2 == dy2 * dx1;
    }
};
# @lc code=end

File: 3625.count-number-of-trapezoids-ii.java
#
# @lc app=leetcode id=3625 lang=java
#
# [3625] Count Number of Trapezoids II
#

# @lc code=start
import java.util.*;

class Solution {
    public int countTrapezoids(int[][] points) {
        int n = points.length;
        Map<String, List<int[]>> slopeToPairs = new HashMap<>();
        // Step 1: For each pair, compute slope as reduced fraction
        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j) {
                int dx = points[j][0] - points[i][0];
                int dy = points[j][1] - points[i][1];
                int g = gcd(dx, dy);
                dx /= g; dy /= g;
                // Normalize direction
                if (dx < 0) { dx = -dx; dy = -dy; }
                else if (dx == 0 && dy < 0) { dy = -dy; }
                String key = dx + "," + dy;
                slopeToPairs.computeIfAbsent(key, k -> new ArrayList<>()).add(new int[]{i, j});
            }
        }
        Set<String> unique = new HashSet<>();
        int res = 0;
        // Step 2: For each group of pairs with same slope, process all disjoint pairs
        for (List<int[]> group : slopeToPairs.values()) {
            int m = group.size();
            for (int a = 0; a < m; ++a) {
                for (int b = a + 1; b < m; ++b) {
                    int[] p1 = group.get(a), p2 = group.get(b);
                    Set<Integer> quad = new HashSet<>(Arrays.asList(p1[0], p1[1], p2[0], p2[1]));
                    if (quad.size() < 4) continue; // not 4 distinct points
                    int[] idxs = quad.stream().mapToInt(x -> x).toArray();
                    // Step 3: Verify convexity
                    if (!isConvex(points, idxs)) continue;
                    int[] sorted = idxs.clone();
                    Arrays.sort(sorted);
                    String sig = sorted[0] + "," + sorted[1] + "," + sorted[2] + "," + sorted[3];
                    if (unique.add(sig)) res++;
                }
            }
        }
        return res;
    }
    // Helper for GCD
    private int gcd(int a, int b) {
        if (b == 0) return Math.abs(a);
        return gcd(b, a % b);
    }
    // Check convexity via cross product
    private boolean isConvex(int[][] pts, int[] idxs) {
        List<int[]> list = new ArrayList<>();
        for (int i : idxs) list.add(pts[i]);
        // Try all circular orderings, find one that is convex
        int perm[] = new int[4];
        for (int i = 0; i < 4; ++i) perm[i] = i;
        do {
            boolean ok = true;
            int prev = 0;
            for (int j = 0; j < 4; ++j) {
                int[] a = list.get(perm[j]);
                int[] b = list.get(perm[(j + 1) % 4]);
                int[] c = list.get(perm[(j + 2) % 4]);
                int cp = cross(a, b, c);
                if (j == 0) prev = cp;
                else if ((long)cp * prev < 0) { ok = false; break; }
            }
            if (ok) return true;
        } while (nextPerm(perm));
        return false;
    }
    private int cross(int[] a, int[] b, int[] c) {
        int x1 = b[0] - a[0], y1 = b[1] - a[1];
        int x2 = c[0] - b[0], y2 = c[1] - b[1];
        return x1 * y2 - y1 * x2;
    }
    // Lexical permutation
    private boolean nextPerm(int[] arr) {
        int i = arr.length - 2;
        while (i >= 0 && arr[i] >= arr[i+1]) i--;
        if (i < 0) return false;
        int j = arr.length - 1;
        while (arr[j] <= arr[i]) j--;
        int t = arr[i]; arr[i] = arr[j]; arr[j] = t;
        Arrays.sort(arr, i+1, arr.length);
        return true;
    }
}
# @lc code=end